logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -319,16 +320,64 @@ private void newCheckpointedHeadpage(int pageNum) throws IOException {
         this.headPage.forceCheckpoint();
     }
 
+    /**
+     * Write a {@link Queueable} element to the queue. This will block until the write succeeds
+     * @param queueable
+     * @return the written sequence number
+     * @throws IOException
+     */
+    public long write(Queueable queueable) throws IOException {
+        return write(queueable, -1);
+    }
+
+    /**
+     * Write a {@link Queueable} element to the queue. This will block until the write succeeds or timeout occurs
+     * @param queueable
+     * @param timeoutMillis maximum time to block before returning. -1 for infinity
+     * @return the written sequence number, or -1 if timed out
+     * @throws IOException
+     */
+    public long write(Queueable queueable, long timeoutMillis)  throws IOException {
+        return write(Collections.singletonList(queueable), timeoutMillis);
+    }
+
     /**
      * write a {@link Queueable} element to the queue. Note that the element will always be written and the queue full
-     * condition will be checked and waited on **after** the write operation.
+     condition will be checked and waited on **after** the write operation.
      *
-     * @param element the {@link Queueable} element to write
-     * @return the written sequence number
+     * @param queueables list of events
+     * @return the maximum written sequence number
+     * @throws IOException
+     */
+    public long write(List<Queueable> queueables) throws IOException {
+        return write(queueables, -1);
+    }
+
+    /**
+     * Write a {@link Queueable} element to the queue. This will block until the write succeeds or a timeout occurs
+     *
+     * @param queueables list of events
+     * @param timeoutMillis maximum time to block before returning. -1 for infinity
+     * @return the maximum written sequence number, or -1 if timed out
      * @throws IOException
      */
-    public long write(Queueable element) throws IOException {
-        byte[] data = element.serialize();
+    public long write(List<Queueable> queueables, long timeoutMillis) throws IOException {
+        if (queueables.isEmpty()) return -1;
+
+        try {
+            timeoutMillis -= optionallyLockForTimeout(timeoutMillis);
```

Comment:
@andrewvc this doesn't seem safe. We should call `optionallyLockForTimeOut` in front of a `try-finally` construct and do whatever we have to do inside that `try`.
This implementation would e.g. block forever on a serialization exception.

That said, I think we can do this easily by simply moving this all below the serialization `for` loop. That would also have the added benefit of not needlessly holding on to the lock while serializing :)

Reply:
Yeah, you're right, I don't know what I was thinking.